File: pipeline/src/lsvoxel/chunkpack/row_to_voxel.py
"""row_to_voxel.bin — whole-dataset, dense-by-row_id lookup from point to its voxel.

This is the integration piece the plan's synthesis pass added: since the 2D and 3D
UMAP fits are independent optimizations (not one fit with an axis dropped), the 2D
minimap can't derive a point's 3D voxel from coordinates alone. Needed for the
minimap's flashlight (2D region -> 3D voxels), crosshair (3D voxel -> 2D point via
coords2d[row_id]), and click-to-teleport.

Layout (8B/record, dense, index == row_id):
  chunk_id:u32  local_voxel_id:u16  reserved:u16
"""
from __future__ import annotations

from pathlib import Path

import numpy as np

ROW_TO_VOXEL_DTYPE = np.dtype(
    [
        ("chunk_id", "<u4"),
        ("local_voxel_id", "<u2"),
        ("reserved", "<u2"),
    ]
)
assert ROW_TO_VOXEL_DTYPE.itemsize == 8, ROW_TO_VOXEL_DTYPE.itemsize


def build_row_to_voxel(n_points: int, assign: dict, out_path: Path) -> None:
    if len(assign["chunk_id"]) != n_points or len(assign["local_voxel_id"]) != n_points:
        raise ValueError("assign arrays must be row_id-ordered, length == n_points")
    recs = np.zeros(n_points, dtype=ROW_TO_VOXEL_DTYPE)
    recs["chunk_id"] = assign["chunk_id"]
    recs["local_voxel_id"] = assign["local_voxel_id"]
    out_path.parent.mkdir(parents=True, exist_ok=True)
    recs.tofile(out_path)


def read_row_to_voxel(path: Path) -> np.ndarray:
    return np.fromfile(path, dtype=ROW_TO_VOXEL_DTYPE)

File: frontend/src/streaming/RowToVoxel.ts
import { fetchArrayBuffer } from "../net/fetchTyped.ts";
import type { Manifest } from "./Manifest.ts";

const RECORD_BYTES = 8;

/**
 * Decoded `row_to_voxel.bin` — whole-dataset, dense by `row_id`, 8B/record:
 *
 *   chunk_id u32 | local_voxel_id u16 | reserved u16   (all little-endian)
 *
 * This is the ONLY bridge from a `row_id` to a place in the 3D world, and it
 * exists precisely because the 2D and 3D UMAP fits are independent
 * optimizations rather than one fit with an axis dropped — there is no
 * coordinate transform between the two spaces, so every 2D↔3D interaction
 * (minimap flashlight, click-to-teleport) has to route through this table.
 * See `pipeline/src/lsvoxel/chunkpack/row_to_voxel.py`, which writes it.
 */
export interface RowToVoxel {
  /** chunkId[row_id] — always an *occupied* chunk (every point lands in one). */
  chunkId: Uint32Array;
  /** localVoxelId[row_id] — 0..voxels_per_chunk^3-1, the atlas tile index too. */
  localVoxelId: Uint16Array;
}

/**
 * Every field in this table happens to be 4-byte-aligned relative to the
 * buffer start (the record is 8 bytes and `chunk_id` sits at offset 0), so
 * unlike `point_index.bin` — whose `local_idx` is at a 2-byte offset and
 * therefore needs a `DataView` — this can be read through a plain
 * `Uint32Array` view: word `2i` is the chunk_id, word `2i+1` packs
 * `local_voxel_id` in its low 16 bits and `reserved` in its high 16.
 */
export function parseRowToVoxel(buffer: ArrayBuffer): RowToVoxel {
  if (buffer.byteLength % RECORD_BYTES !== 0) {
    throw new Error(
      `row_to_voxel.bin: size ${buffer.byteLength}B isn't a multiple of ${RECORD_BYTES}B`,
    );
  }
  const words = new Uint32Array(buffer);
  const n = words.length / 2;
  const chunkId = new Uint32Array(n);
  const localVoxelId = new Uint16Array(n);
  for (let i = 0; i < n; i++) {
    chunkId[i] = words[i * 2];
    localVoxelId[i] = words[i * 2 + 1] & 0xffff;
  }
  return { chunkId, localVoxelId };
}

/** Fetches and parses `row_to_voxel.bin` for a loaded chunk-pack manifest. */
export async function loadRowToVoxel(
  manifest: Manifest,
  signal?: AbortSignal,
): Promise<RowToVoxel> {
  const buffer = await fetchArrayBuffer(manifest.url(manifest.raw.row_to_voxel.path), signal);
  const table = parseRowToVoxel(buffer);
  if (table.chunkId.length !== manifest.totalPoints) {
    throw new Error(
      `row_to_voxel.bin has ${table.chunkId.length} records, manifest says ${manifest.totalPoints} points`,
    );
  }
  return table;
}
